File: mockspresso-basic-plugins/src/main/java/com/episode6/hackit/mockspresso/basic/plugin/simple/SimpleInjectionConfig.java
package com.episode6.hackit.mockspresso.basic.plugin.simple;

import com.episode6.hackit.mockspresso.api.InjectionConfig;
import com.episode6.hackit.mockspresso.reflect.TypeToken;

import org.jetbrains.annotations.Nullable;
import java.lang.annotation.Annotation;
import java.lang.reflect.Constructor;
import java.util.Collections;
import java.util.List;

/**
 * A very simple implementation of InjectionConfig. Chooses the constructor with the least number of arguments,
 * and provides no injectable field annotations.
 */
public class SimpleInjectionConfig implements InjectionConfig {

  private final ConstructorSelector mConstructorSelector = new SimpleConstructorSelector();

  @Override
  public ConstructorSelector provideConstructorSelector() {
    return mConstructorSelector;
  }

  @Override
  public List<Class<? extends Annotation>> provideInjectableFieldAnnotations() {
    return Collections.emptyList();
  }

  @Override
  public List<Class<? extends Annotation>> provideInjectableMethodAnnotations() {
    return Collections.emptyList();
  }

  private static class SimpleConstructorSelector implements ConstructorSelector {

    @SuppressWarnings("unchecked")
    @Override
    public @Nullable <T> Constructor<T> chooseConstructor(TypeToken<T> typeToken) {
      Constructor<T> found = null;
      for (Constructor<?> constructor : typeToken.getRawType().getDeclaredConstructors()) {
        if (found == null || constructor.getParameterCount() < found.getParameterCount()) {
          found = (Constructor<T>) constructor;
        }
      }
      return found;
    }
  }
}

File: mockspresso-basic-plugins/src/main/java/com/episode6/hackit/mockspresso/basic/plugin/simple/SimpleInjectionConfig.kt
package com.episode6.hackit.mockspresso.basic.plugin.simple

import com.episode6.hackit.mockspresso.api.InjectionConfig
import com.episode6.hackit.mockspresso.reflect.TypeToken
import java.lang.reflect.Constructor

/**
 * A very simple implementation of InjectionConfig. Chooses the constructor with the least number of arguments,
 * and provides no injectable field annotations.
 */
internal class SimpleInjectionConfig : InjectionConfig {
  override fun provideInjectableFieldAnnotations(): List<Class<out Annotation>> = emptyList()
  override fun provideInjectableMethodAnnotations(): List<Class<out Annotation>> = emptyList()

  @Suppress("UNCHECKED_CAST")
  override fun <T : Any> chooseConstructor(typeToken: TypeToken<T>): Constructor<T>? = typeToken.rawType
      .declaredConstructors
      .minByOrNull { it.parameterCount } as Constructor<T>?
}
